### PR TITLE
[VPlan] Document current status.

### PR DESCRIPTION
**Pull request by fhahn**
Update VectorizationPlan.rst to include a section about the current status of VPlan and its use in LoopVectorize, based on "VPlan: Status Update and Future Roadmap", LLVM Developers’ Meeting 2023, https://www.youtube.com/watch?v=SzGP4PgMuLE

Preview a rendered version here:
https://gist.github.com/fhahn/c4a6d7121b86904ab5e0fdf30ab54940/raw/45bea52de83f4ab500a534d2627d8d6a101cb390/VectorizationPlan-CurrentStatus.pdf

### CODE REVIEW

File: llvm/docs/VectorizationPlan.rst [hunk 1]
```diff
@@ -19,6 +19,49 @@ candidates. It serves for both optimizing candidates including estimating their
...
 
+Current Status
+==============
+VPlan is currently used to drive code-generation in LoopVectorize. VPlans are
+constructed after all cost-based and most legality-related decisions have been
+taken. As def-use chains between recipes are now fully modeled in VPlan,
+VPlan-based analyses and transformations are used to simplify and modularize
+the vectorization process [10]_. Those include transformations to
+
+1. Legalize the initial VPlan, e.g. by introducing specializedrecipes for
```
Comment (michaelmaitland) [inline llvm/docs/VectorizationPlan.rst:30]: specializedrecipes -> specialized recipes

File: llvm/docs/VectorizationPlan.rst [hunk 2]
```diff
@@ -19,6 +19,49 @@ candidates. It serves for both optimizing candidates including estimating their
 cost reliably, and for performing their final translation into IR. This
 facilitates dealing with multiple vectorization candidates.
 
+Current Status
+==============
+VPlan is currently used to drive code-generation in LoopVectorize. VPlans are
+constructed after all cost-based and most legality-related decisions have been
```
Comment (michaelmaitland) [inline llvm/docs/VectorizationPlan.rst:25]: > VPlans are constructed after all cost-based ... decisions have been taken

It is my understanding that we build VPlans and then we determine the cost of each VPlan we have built to find the best VPlan. I haven't had a chance to work in LV for a few months now so I am not so familiar with recent changes. Did we move away from this, or are you describing a different kind of cost decisions?

File: llvm/docs/conf.py [hunk 3]
```diff
@@ -222,6 +222,7 @@
 # If false, no module index is generated.
 # latex_domain_indices = True
 
+numfig = True
```
Comment (michaelmaitland) [inline llvm/docs/conf.py:225]: nit: can you add a comment describing the purpose of enabling this?

File: llvm/docs/VectorizationPlan.rst [hunk 4]
```diff
@@ -19,6 +19,49 @@ candidates. It serves for both optimizing candidates including estimating their
...
+
+.. _fig-vplan-scope:
+.. figure:: ./vplan-scope.png
+   :width: 800 px
+
+   Scope modeled in VPlan in 2024
+
+
 High-level Design
 =================
```
Comment (ayalz) [inline llvm/docs/VectorizationPlan.rst:67]: While we're here, worth also updating _the following_:

...
:VPInstruction:
  A VPInstruction is [both a VPRecipe and a VPUser -> _a recipe characterized by a single opcode and optional flags, free of ingredients or other meta-data_]. [_remove:_ It models a single
  VPlan-level instruction to be generated if the VPlan is executed, including
  its opcode and possibly additional characteristics. It is the basis for
  writing instruction-level analyses and optimizations in VPlan as creating,
  replacing or moving VPInstructions record both def-use and scheduling
  decisions.] VPInstructions also extend LLVM IR's opcodes with idiomatic
  operations that enrich the Vectorizer's semantics.

BTW, it seems that VPRecipesBase::mayReadFromMemory() and mayWriteToMemory() are missing a `case VPInstructionSC:`, similar to VPRecipeBase::mayHaveSideEffects()?
...

The Planning Process and VPlan Roadmap
======================================
Transforming the Loop Vectorizer to use VPlan follows a staged approach. First,
VPlan _[is -> was only]_ used to record the final vectorization decisions, and to execute them:
the Hierarchical CFG models the planned control-flow, and Recipes capture
decisions taken inside basic-blocks. _[Next, VPlan will be -> Currently, VPlan is]_ used also as the basis
for taking these decisions, effectively turning them into a series of
VPlan-to-VPlan algorithms. Finally, VPlan will support the planning process
itself including cost-based analyses for making these decisions, to fully
support compositional and iterative decision making.

**Review by ayalz — APPROVED** [review]
Thanks for following-up on this! LGTM, adding minor comments.

File: llvm/docs/VectorizationPlan.rst [hunk 5]
```diff
@@ -248,3 +291,6 @@ References
 
 .. [9] "Extending LoopVectorizer: OpenMP4.5 SIMD and Outer Loop
     Auto-Vectorization", Intel Vectorizer Team, LLVM Developers' Meeting 2016.
+
+.. [10] "VPlan: Status Update and Future Roadmap", Florian Hahn, LLVM
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/docs/VectorizationPlan.rst:292]: Missed originally, should be addressed in 309a881dccb82bf1f101cf138bee3e7400968ed8

File: llvm/docs/conf.py (hunk 3 above)
Comment (fhahn) [reply llvm/docs/conf.py:225]: Missed originally, should be addressed in 309a881dccb82bf1f101cf138bee3e7400968ed8

**Pull request merged**